crates/typst/src/model/introspect.rs

```diff
@@ -294,6 +291,68 @@ impl Introspector {
                 }
                 list
             }
+            Selector::And(selectors) => {
+                let all_matches: Vec<_> =
+                    selectors.iter().map(|sel| self.query(sel)).collect();
+
+                let mut result = EcoVec::new();
+
+                if let Some(first) = all_matches.first() {
+                    let other_lists = &all_matches[1..];
+
+                    let is_element =
+                        |candidate, other_list: &EcoVec<Prehashed<Content>>| {
+                            matches!(
+                                other_list.binary_search_by_key(
+                                    &self.index(candidate),
+                                    |elem| { self.index(elem) }
+                                ),
+                                Ok(_)
+                            )
+                        };
+
+                    for candidate in first.iter() {
```

Comment:
If `n` is the max number of matches in a subquery and `k` the number of subqueries, this should run in `n * k * log n`. However, I believe `n * k` would be possible because all result lists are in the same order. Basically, we would store in all_matches peekable iterators instead of vectors and then advance these in lockstep:

Advance the first one, then advance all other ones while their first element has a smaller or equal index than the one that the first yielded. And if all iterators yielded the an element during this procedure, keep it.